// @moq/lite/stream.js
import * as Varint from "./varint.js";
const MAX_U31 = 2 ** 31 - 1;
const MAX_READ_SIZE = 1024 * 1024 * 64; // don't allocate more than 64MB for a message
export class Stream {
    reader;
    writer;
    constructor(props) {
        this.writer = new Writer(props.writable);
        this.reader = new Reader(props.readable);
    }
    static async accept(quic) {
        for (;;) {
            const reader = quic.incomingBidirectionalStreams.getReader();
            const next = await reader.read();
            reader.releaseLock();
            if (next.done)
                return;
            return new Stream(next.value);
        }
    }
    static async open(quic, priority) {
        return new Stream(await quic.createBidirectionalStream({ sendOrder: priority }));
    }
    close() {
        this.writer.close();
        this.reader.stop(new Error("cancel"));
    }
    abort(reason) {
        this.writer.reset(reason);
        this.reader.stop(reason);
    }
}
// Reader wraps a stream and provides convience methods for reading pieces from a stream
// Unfortunately we can't use a BYOB reader because it's not supported with WebTransport+WebWorkers yet.
export class Reader {
    #buffer;
    #stream; // if undefined, the buffer is consumed then EOF
    #reader;
    constructor(stream, buffer) {
        this.#buffer = buffer ?? new Uint8Array();
        this.#stream = stream;
        this.#reader = this.#stream?.getReader();
    }
    // Adds more data to the buffer, returning true if more data was added.
    async #fill() {
        if (!this.#reader) {
            return false;
        }
        const result = await this.#reader.read();
        if (result.done) {
            return false;
        }
        if (result.value.byteLength === 0) {
            throw new Error("unexpected empty chunk");
        }
        const buffer = new Uint8Array(result.value);
        if (this.#buffer.byteLength === 0) {
            this.#buffer = buffer;
        }
        else {
            const temp = new Uint8Array(this.#buffer.byteLength + buffer.byteLength);
            temp.set(this.#buffer);
            temp.set(buffer, this.#buffer.byteLength);
            this.#buffer = temp;
        }
        return true;
    }
    // Add more data to the buffer until it's at least size bytes.
    async #fillTo(size) {
        if (size > MAX_READ_SIZE) {
            throw new Error(`read size ${size} exceeds max size ${MAX_READ_SIZE}`);
        }
        while (this.#buffer.byteLength < size) {
            if (!(await this.#fill())) {
                throw new Error("unexpected end of stream");
            }
        }
    }
    // Consumes the first size bytes of the buffer.
    #slice(size) {
        const result = new Uint8Array(this.#buffer.buffer, this.#buffer.byteOffset, size);
        this.#buffer = new Uint8Array(this.#buffer.buffer, this.#buffer.byteOffset + size, this.#buffer.byteLength - size);
        return result;
    }
    async read(size) {
        if (size === 0)
            return new Uint8Array();
        await this.#fillTo(size);
        return this.#slice(size);
    }
    async readAll() {
        while (await this.#fill()) {
            // keep going
        }
        return this.#slice(this.#buffer.byteLength);
    }
    async string() {
        const length = await this.u53();
        const buffer = await this.read(length);
        return new TextDecoder().decode(buffer);
    }
    async bool() {
        const v = await this.u8();
        if (v === 0)
            return false;
        if (v === 1)
            return true;
        throw new Error("invalid bool value");
    }
    async u8() {
        await this.#fillTo(1);
        return this.#slice(1)[0];
    }
    async u16() {
        await this.#fillTo(2);
        const view = new DataView(this.#buffer.buffer, this.#buffer.byteOffset, 2);
        const result = view.getUint16(0);
        this.#slice(2);
        return result;
    }
    // Returns a Number using 53-bits, the max Javascript can use for integer math
    async u53() {
        const v = await this.u62();
        if (v > Varint.MAX_U53) {
            throw new Error("value larger than 53-bits; use v62 instead");
        }
        return Number(v);
    }
    // NOTE: Returns a bigint instead of a number since it may be larger than 53-bits
    async u62() {
        await this.#fillTo(1);
        const size = (this.#buffer[0] & 0xc0) >> 6;
        if (size === 0) {
            const first = this.#slice(1)[0];
            return BigInt(first) & 0x3fn;
        }
        if (size === 1) {
            await this.#fillTo(2);
            const slice = this.#slice(2);
            const view = new DataView(slice.buffer, slice.byteOffset, slice.byteLength);
            return BigInt(view.getUint16(0)) & 0x3fffn;
        }
        if (size === 2) {
            await this.#fillTo(4);
            const slice = this.#slice(4);
            const view = new DataView(slice.buffer, slice.byteOffset, slice.byteLength);
            return BigInt(view.getUint32(0)) & 0x3fffffffn;
        }
        await this.#fillTo(8);
        const slice = this.#slice(8);
        const view = new DataView(slice.buffer, slice.byteOffset, slice.byteLength);
        return view.getBigUint64(0) & 0x3fffffffffffffffn;
    }
    // Returns false if there is more data to read, blocking if it hasn't been received yet.
    async done() {
        if (this.#buffer.byteLength > 0)
            return false;
        return !(await this.#fill());
    }
    stop(reason) {
        this.#reader?.cancel(reason).catch(() => void 0);
    }
    get closed() {
        return this.#reader?.closed ?? Promise.resolve();
    }
}
// Writer wraps a stream and writes chunks of data
export class Writer {
    #writer;
    #stream;
    // Scratch buffer for writing varints.
    // Fixed at 8 bytes.
    #scratch;
    constructor(stream) {
        this.#stream = stream;
        this.#scratch = new ArrayBuffer(8);
        this.#writer = this.#stream.getWriter();
    }
    async bool(v) {
        await this.write(setUint8(this.#scratch, v ? 1 : 0));
    }
    async u8(v) {
        await this.write(setUint8(this.#scratch, v));
    }
    async u16(v) {
        await this.write(setUint16(this.#scratch, v));
    }
    async i32(v) {
        if (Math.abs(v) > MAX_U31) {
            throw new Error(`overflow, value larger than 32-bits: ${v.toString()}`);
        }
        // We don't use a VarInt, so it always takes 4 bytes.
        // This could be improved but nothing is standardized yet.
        await this.write(setInt32(this.#scratch, v));
    }
    async u53(v) {
        if (v > Varint.MAX_U53) {
            throw new Error(`overflow, value larger than 53-bits: ${v.toString()}`);
        }
        await this.write(Varint.encodeTo(this.#scratch, v));
    }
    async u62(v) {
        await this.write(Varint.encodeTo(this.#scratch, v));
    }
    async write(v) {
        await this.#writer.write(v);
    }
    async string(str) {
        const data = new TextEncoder().encode(str);
        await this.u53(data.byteLength);
        await this.write(data);
    }
    close() {
        this.#writer.close().catch(() => void 0);
    }
    get closed() {
        return this.#writer.closed;
    }
    reset(reason) {
        this.#writer.abort(reason).catch(() => void 0);
    }
    static async open(quic) {
        const writable = (await quic.createUnidirectionalStream());
        return new Writer(writable);
    }
}
function setUint8(dst, v) {
    const buffer = new Uint8Array(dst, 0, 1);
    buffer[0] = v;
    return buffer;
}
function setUint16(dst, v) {
    const view = new DataView(dst, 0, 2);
    view.setUint16(0, v);
    return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
}
function setInt32(dst, v) {
    const view = new DataView(dst, 0, 4);
    view.setInt32(0, v);
    return new Uint8Array(view.buffer, view.byteOffset, view.byteLength);
}
// Returns the next stream from the connection
export class Readers {
    #reader;
    constructor(quic) {
        this.#reader = quic.incomingUnidirectionalStreams.getReader();
    }
    async next() {
        const next = await this.#reader.read();
        if (next.done)
            return;
        return new Reader(next.value);
    }
    close() {
        this.#reader.cancel();
    }
}
