// @moq/lite/ietf/setup.js
import { unreachable } from "../util/error.js";
import * as Message from "./message.js";
import { Parameters } from "./parameters.js";
import { Version } from "./version.js";
const MAX_VERSIONS = 128;
export class ClientSetup {
    static id = 0x20;
    versions;
    parameters;
    constructor({ versions, parameters = new Parameters() }) {
        this.versions = versions;
        this.parameters = parameters;
    }
    async #encode(w, version) {
        if (version === Version.DRAFT_15 || version === Version.DRAFT_16) {
            // Draft15+: no versions list, just parameters
            await this.parameters.encode(w, version);
        }
        else if (version === Version.DRAFT_14) {
            await w.u53(this.versions.length);
            for (const v of this.versions) {
                await w.u53(v);
            }
            await this.parameters.encode(w, version);
        }
        else {
            unreachable(version);
        }
    }
    async encode(w, version) {
        return Message.encode(w, (mw) => this.#encode(mw, version));
    }
    static async #decode(r, version) {
        if (version === Version.DRAFT_15 || version === Version.DRAFT_16) {
            // Draft15+: no versions list, just parameters
            const parameters = await Parameters.decode(r, version);
            return new ClientSetup({ versions: [version], parameters });
        }
        else if (version === Version.DRAFT_14) {
            // Number of supported versions
            const numVersions = await r.u53();
            if (numVersions > MAX_VERSIONS) {
                throw new Error(`too many versions: ${numVersions}`);
            }
            const supportedVersions = [];
            for (let i = 0; i < numVersions; i++) {
                const v = await r.u53();
                supportedVersions.push(v);
            }
            const parameters = await Parameters.decode(r, version);
            return new ClientSetup({ versions: supportedVersions, parameters });
        }
        else {
            unreachable(version);
        }
    }
    static async decode(r, version) {
        return Message.decode(r, (mr) => ClientSetup.#decode(mr, version));
    }
}
export class ServerSetup {
    static id = 0x21;
    version;
    parameters;
    constructor({ version, parameters = new Parameters() }) {
        this.version = version;
        this.parameters = parameters;
    }
    async #encode(w, version) {
        if (version === Version.DRAFT_15 || version === Version.DRAFT_16) {
            // Draft15+: no version field, just parameters
            await this.parameters.encode(w, version);
        }
        else if (version === Version.DRAFT_14) {
            await w.u53(this.version);
            await this.parameters.encode(w, version);
        }
        else {
            unreachable(version);
        }
    }
    async encode(w, version) {
        return Message.encode(w, (mw) => this.#encode(mw, version));
    }
    static async #decode(r, version) {
        if (version === Version.DRAFT_15 || version === Version.DRAFT_16) {
            // Draft15+: no version field, just parameters
            const parameters = await Parameters.decode(r, version);
            return new ServerSetup({ version, parameters });
        }
        else if (version === Version.DRAFT_14) {
            const selectedVersion = await r.u53();
            const parameters = await Parameters.decode(r, version);
            return new ServerSetup({ version: selectedVersion, parameters });
        }
        else {
            unreachable(version);
        }
    }
    static async decode(r, version) {
        return Message.decode(r, (mr) => ServerSetup.#decode(mr, version));
    }
}
