// @moq/lite/lite/fetch.js
import * as Path from "../path.js";
import { unreachable } from "../util/error.js";
import * as Message from "./message.js";
import { Version } from "./version.js";
function guardDraft03(version) {
    switch (version) {
        case Version.DRAFT_03:
            break;
        case Version.DRAFT_01:
        case Version.DRAFT_02:
            throw new Error("fetch not supported for this version");
        default:
            unreachable(version);
    }
}
export class Fetch {
    broadcast;
    track;
    priority;
    group;
    constructor(broadcast, track, priority, group) {
        this.broadcast = broadcast;
        this.track = track;
        this.priority = priority;
        this.group = group;
    }
    async #encode(w) {
        await w.string(this.broadcast);
        await w.string(this.track);
        await w.u8(this.priority);
        await w.u53(this.group);
    }
    static async #decode(r) {
        const broadcast = Path.from(await r.string());
        const track = await r.string();
        const priority = await r.u8();
        const group = await r.u53();
        return new Fetch(broadcast, track, priority, group);
    }
    async encode(w, version) {
        guardDraft03(version);
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, version) {
        guardDraft03(version);
        return Message.decode(r, Fetch.#decode);
    }
}
