// @moq/lite/lite/group.js
import * as Message from "./message.js";
export class Group {
    subscribe;
    sequence;
    constructor(subscribe, sequence) {
        this.subscribe = subscribe;
        this.sequence = sequence;
    }
    async #encode(w) {
        await w.u62(this.subscribe);
        await w.u53(this.sequence);
    }
    static async #decode(r) {
        return new Group(await r.u62(), await r.u53());
    }
    async encode(w) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r) {
        return Message.decode(r, Group.#decode);
    }
    static async decodeMaybe(r) {
        return Message.decodeMaybe(r, Group.#decode);
    }
}
export class Frame {
    payload;
    constructor(payload) {
        this.payload = payload;
    }
    async #encode(w) {
        await w.write(this.payload);
    }
    static async #decode(r) {
        const payload = await r.readAll();
        return new Frame(payload);
    }
    async encode(w) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r) {
        return Message.decode(r, Frame.#decode);
    }
}
