// @moq/lite/lite/probe.js
import { unreachable } from "../util/error.js";
import * as Message from "./message.js";
import { Version } from "./version.js";
function guardDraft03(version) {
    switch (version) {
        case Version.DRAFT_03:
            break;
        case Version.DRAFT_01:
        case Version.DRAFT_02:
            throw new Error("probe not supported for this version");
        default:
            unreachable(version);
    }
}
export class Probe {
    bitrate;
    constructor(bitrate) {
        this.bitrate = bitrate;
    }
    async #encode(w) {
        await w.u53(this.bitrate);
    }
    static async #decode(r) {
        return new Probe(await r.u53());
    }
    async encode(w, version) {
        guardDraft03(version);
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, version) {
        guardDraft03(version);
        return Message.decode(r, Probe.#decode);
    }
    static async decodeMaybe(r, version) {
        guardDraft03(version);
        return Message.decodeMaybe(r, Probe.#decode);
    }
}
