// @moq/lite/lite/subscribe.js
import * as Path from "../path.js";
import { unreachable } from "../util/error.js";
import * as Message from "./message.js";
import { Version } from "./version.js";
export class SubscribeUpdate {
    priority;
    ordered;
    maxLatency;
    startGroup;
    endGroup;
    constructor(props) {
        this.priority = props.priority;
        this.ordered = props.ordered ?? true;
        this.maxLatency = props.maxLatency ?? 0;
        this.startGroup = props.startGroup;
        this.endGroup = props.endGroup;
    }
    async #encode(w, version) {
        switch (version) {
            case Version.DRAFT_03:
                await w.u8(this.priority);
                await w.bool(this.ordered);
                await w.u53(this.maxLatency);
                await w.u53(this.startGroup !== undefined ? this.startGroup + 1 : 0);
                await w.u53(this.endGroup !== undefined ? this.endGroup + 1 : 0);
                break;
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                await w.u8(this.priority);
                break;
            default:
                unreachable(version);
        }
    }
    static async #decode(r, version) {
        switch (version) {
            case Version.DRAFT_03: {
                const priority = await r.u8();
                const ordered = await r.bool();
                const maxLatency = await r.u53();
                const startGroup = await r.u53();
                const endGroup = await r.u53();
                return new SubscribeUpdate({
                    priority,
                    ordered,
                    maxLatency,
                    startGroup: startGroup > 0 ? startGroup - 1 : undefined,
                    endGroup: endGroup > 0 ? endGroup - 1 : undefined,
                });
            }
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                return new SubscribeUpdate({ priority: await r.u8() });
            default:
                unreachable(version);
        }
    }
    async encode(w, version) {
        return Message.encode(w, (w) => this.#encode(w, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (r) => SubscribeUpdate.#decode(r, version));
    }
    static async decodeMaybe(r, version) {
        return Message.decodeMaybe(r, (r) => SubscribeUpdate.#decode(r, version));
    }
}
export class Subscribe {
    id;
    broadcast;
    track;
    priority;
    ordered;
    maxLatency;
    startGroup;
    endGroup;
    constructor(props) {
        this.id = props.id;
        this.broadcast = props.broadcast;
        this.track = props.track;
        this.priority = props.priority;
        this.ordered = props.ordered ?? false;
        this.maxLatency = props.maxLatency ?? 0;
        this.startGroup = props.startGroup;
        this.endGroup = props.endGroup;
    }
    async #encode(w, version) {
        await w.u62(this.id);
        await w.string(this.broadcast);
        await w.string(this.track);
        await w.u8(this.priority);
        switch (version) {
            case Version.DRAFT_03:
                await w.bool(this.ordered);
                await w.u53(this.maxLatency);
                await w.u53(this.startGroup !== undefined ? this.startGroup + 1 : 0);
                await w.u53(this.endGroup !== undefined ? this.endGroup + 1 : 0);
                break;
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                break;
            default:
                unreachable(version);
        }
    }
    static async #decode(r, version) {
        const id = await r.u62();
        const broadcast = Path.from(await r.string());
        const track = await r.string();
        const priority = await r.u8();
        switch (version) {
            case Version.DRAFT_03: {
                const ordered = await r.bool();
                const maxLatency = await r.u53();
                const startGroup = await r.u53();
                const endGroup = await r.u53();
                return new Subscribe({
                    id,
                    broadcast,
                    track,
                    priority,
                    ordered,
                    maxLatency,
                    startGroup: startGroup > 0 ? startGroup - 1 : undefined,
                    endGroup: endGroup > 0 ? endGroup - 1 : undefined,
                });
            }
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                return new Subscribe({ id, broadcast, track, priority });
            default:
                unreachable(version);
        }
    }
    async encode(w, version) {
        return Message.encode(w, (w) => this.#encode(w, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (r) => Subscribe.#decode(r, version));
    }
}
export class SubscribeOk {
    priority;
    ordered;
    maxLatency;
    startGroup;
    endGroup;
    constructor({ priority = 0, ordered = true, maxLatency = 0, startGroup = undefined, endGroup = undefined, }) {
        this.priority = priority;
        this.ordered = ordered;
        this.maxLatency = maxLatency;
        this.startGroup = startGroup;
        this.endGroup = endGroup;
    }
    async #encode(w, version) {
        switch (version) {
            case Version.DRAFT_03:
                await w.u8(this.priority);
                await w.bool(this.ordered);
                await w.u53(this.maxLatency);
                await w.u53(this.startGroup !== undefined ? this.startGroup + 1 : 0);
                await w.u53(this.endGroup !== undefined ? this.endGroup + 1 : 0);
                break;
            case Version.DRAFT_02:
                // noop
                break;
            case Version.DRAFT_01:
                await w.u8(this.priority ?? 0);
                break;
            default:
                unreachable(version);
        }
    }
    static async #decode(version, r) {
        let priority;
        let ordered;
        let maxLatency;
        let startGroup;
        let endGroup;
        switch (version) {
            case Version.DRAFT_03:
                priority = await r.u8();
                ordered = await r.bool();
                maxLatency = await r.u53();
                startGroup = await r.u53();
                endGroup = await r.u53();
                break;
            case Version.DRAFT_02:
                // noop
                break;
            case Version.DRAFT_01:
                priority = await r.u8();
                break;
            default:
                unreachable(version);
        }
        return new SubscribeOk({
            priority,
            ordered,
            maxLatency,
            startGroup: startGroup !== undefined && startGroup > 0 ? startGroup - 1 : undefined,
            endGroup: endGroup !== undefined && endGroup > 0 ? endGroup - 1 : undefined,
        });
    }
    async encode(w, version) {
        return Message.encode(w, (w) => this.#encode(w, version));
    }
    static async decode(r, version) {
        return Message.decode(r, SubscribeOk.#decode.bind(SubscribeOk, version));
    }
}
/// Indicates that one or more groups have been dropped.
///
/// Draft03 only.
export class SubscribeDrop {
    start;
    end;
    error;
    constructor(props) {
        this.start = props.start;
        this.end = props.end;
        this.error = props.error;
    }
    async #encode(w) {
        await w.u53(this.start);
        await w.u53(this.end);
        await w.u53(this.error);
    }
    static async #decode(r) {
        return new SubscribeDrop({ start: await r.u53(), end: await r.u53(), error: await r.u53() });
    }
    async encode(w) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r) {
        return Message.decode(r, SubscribeDrop.#decode);
    }
}
export async function encodeSubscribeResponse(w, resp, version) {
    switch (version) {
        case Version.DRAFT_03:
            if ("ok" in resp) {
                await w.u53(0x0);
                await resp.ok.encode(w, version);
            }
            else {
                await w.u53(0x1);
                await resp.drop.encode(w);
            }
            break;
        case Version.DRAFT_01:
        case Version.DRAFT_02:
            if ("ok" in resp) {
                await resp.ok.encode(w, version);
            }
            else {
                throw new Error("subscribe drop not supported for this version");
            }
            break;
        default:
            unreachable(version);
    }
}
export async function decodeSubscribeResponse(r, version) {
    switch (version) {
        case Version.DRAFT_03: {
            const typ = await r.u53();
            switch (typ) {
                case 0x0:
                    return { ok: await SubscribeOk.decode(r, version) };
                case 0x1:
                    return { drop: await SubscribeDrop.decode(r) };
                default:
                    throw new Error(`unknown subscribe response type: ${typ}`);
            }
        }
        case Version.DRAFT_01:
        case Version.DRAFT_02:
            return { ok: await SubscribeOk.decode(r, version) };
        default:
            unreachable(version);
    }
}
