// @moq/lite/ietf/connection.js
import * as Path from "../path.js";
import { Readers, Stream } from "../stream.js";
import { unreachable } from "../util/index.js";
import * as Control from "./control.js";
import { Fetch, FetchCancel, FetchError, FetchOk } from "./fetch.js";
import { GoAway } from "./goaway.js";
import { Group } from "./object.js";
import { Publish, PublishDone, PublishError, PublishOk } from "./publish.js";
import { PublishNamespace, PublishNamespaceCancel, PublishNamespaceDone, PublishNamespaceError, PublishNamespaceOk, } from "./publish_namespace.js";
import { Publisher } from "./publisher.js";
import { MaxRequestId, RequestError, RequestOk, RequestsBlocked } from "./request.js";
import * as Setup from "./setup.js";
import { Subscribe, SubscribeError, SubscribeOk, Unsubscribe } from "./subscribe.js";
import { SubscribeNamespace, SubscribeNamespaceError, SubscribeNamespaceOk, UnsubscribeNamespace, } from "./subscribe_namespace.js";
import { Subscriber } from "./subscriber.js";
import { TrackStatus, TrackStatusRequest } from "./track.js";
import { Version } from "./version.js";
/**
 * Represents a connection to a MoQ server using moq-transport protocol.
 *
 * @public
 */
export class Connection {
    // The URL of the connection.
    url;
    // The established WebTransport session.
    #quic;
    // The single bidirectional control stream for control messages
    #control;
    // Module for contributing tracks.
    #publisher;
    // Module for distributing tracks.
    #subscriber;
    // Just to avoid logging when `close()` is called.
    #closed = false;
    /**
     * Creates a new Connection instance.
     * @param url - The URL of the connection
     * @param quic - The WebTransport session
     * @param controlStream - The control stream
     * @param maxRequestId - The initial max request ID
     * @param version - The negotiated protocol version
     *
     * @internal
     */
    constructor({ url, quic, control, maxRequestId, version, }) {
        this.url = url;
        this.#quic = quic;
        this.#control = new Control.Stream({ stream: control, maxRequestId, version });
        this.#quic.closed.finally(() => {
            this.#control.close();
        });
        this.#publisher = new Publisher({ quic: this.#quic, control: this.#control });
        this.#subscriber = new Subscriber({ control: this.#control, quic: this.#quic });
        void this.#run();
    }
    /**
     * Closes the connection.
     */
    close() {
        if (this.#closed)
            return;
        this.#closed = true;
        try {
            this.#quic.close();
        }
        catch {
            // ignore
        }
    }
    async #run() {
        const tasks = [this.#runControlStream(), this.#runObjectStreams()];
        // v16: accept bidi streams for SUBSCRIBE_NAMESPACE
        if (this.#control.version === Version.DRAFT_16) {
            tasks.push(this.#runBidiStreams());
        }
        try {
            await Promise.all(tasks);
        }
        catch (err) {
            if (!this.#closed) {
                console.error("fatal error running connection", err);
            }
        }
        finally {
            this.close();
        }
    }
    /**
     * Publishes a broadcast to the connection.
     * @param name - The broadcast path to publish
     * @param broadcast - The broadcast to publish
     */
    publish(path, broadcast) {
        this.#publisher.publish(path, broadcast);
    }
    /**
     * Gets an announced reader for the specified prefix.
     * @param prefix - The prefix for announcements
     * @returns An AnnounceConsumer instance
     */
    announced(prefix = Path.empty()) {
        return this.#subscriber.announced(prefix);
    }
    /**
     * Consumes a broadcast from the connection.
     *
     * @remarks
     * If the broadcast is not found, a "not found" error will be thrown when requesting any tracks.
     *
     * @param broadcast - The path of the broadcast to consume
     * @returns A Broadcast instance
     */
    consume(broadcast) {
        return this.#subscriber.consume(broadcast);
    }
    /**
     * Handles control messages on the single bidirectional control stream.
     */
    async #runControlStream() {
        for (;;) {
            try {
                const msg = await this.#control.read();
                // Route control messages to appropriate handlers based on type
                // Messages sent by Subscriber, received by Publisher:
                if (msg instanceof Subscribe) {
                    await this.#publisher.handleSubscribe(msg);
                }
                else if (msg instanceof Unsubscribe) {
                    await this.#publisher.handleUnsubscribe(msg);
                }
                else if (msg instanceof TrackStatusRequest) {
                    await this.#publisher.handleTrackStatusRequest(msg);
                }
                else if (msg instanceof PublishNamespaceOk) {
                    await this.#publisher.handlePublishNamespaceOk(msg);
                }
                else if (msg instanceof PublishNamespaceError) {
                    await this.#publisher.handlePublishNamespaceError(msg);
                }
                else if (msg instanceof PublishNamespaceCancel) {
                    await this.#publisher.handlePublishNamespaceCancel(msg);
                }
                else if (msg instanceof PublishNamespace) {
                    await this.#subscriber.handlePublishNamespace(msg);
                }
                else if (msg instanceof PublishNamespaceDone) {
                    await this.#subscriber.handlePublishNamespaceDone(msg);
                }
                else if (msg instanceof SubscribeOk) {
                    await this.#subscriber.handleSubscribeOk(msg);
                }
                else if (msg instanceof SubscribeError) {
                    await this.#subscriber.handleSubscribeError(msg);
                }
                else if (msg instanceof PublishDone) {
                    await this.#subscriber.handlePublishDone(msg);
                }
                else if (msg instanceof TrackStatus) {
                    await this.#subscriber.handleTrackStatus(msg);
                }
                else if (msg instanceof GoAway) {
                    await this.#handleGoAway(msg);
                }
                else if (msg instanceof Setup.ClientSetup) {
                    await this.#handleClientSetup(msg);
                }
                else if (msg instanceof Setup.ServerSetup) {
                    await this.#handleServerSetup(msg);
                }
                else if (msg instanceof SubscribeNamespace) {
                    await this.#publisher.handleSubscribeNamespace(msg);
                }
                else if (msg instanceof SubscribeNamespaceOk) {
                    await this.#subscriber.handleSubscribeNamespaceOk(msg);
                }
                else if (msg instanceof SubscribeNamespaceError) {
                    await this.#subscriber.handleSubscribeNamespaceError(msg);
                }
                else if (msg instanceof UnsubscribeNamespace) {
                    await this.#publisher.handleUnsubscribeNamespace(msg);
                }
                else if (msg instanceof Publish) {
                    await this.#subscriber.handlePublish(msg);
                }
                else if (msg instanceof PublishOk) {
                    throw new Error("PUBLISH_OK messages are not supported");
                }
                else if (msg instanceof PublishError) {
                    throw new Error("PUBLISH_ERROR messages are not supported");
                }
                else if (msg instanceof Fetch) {
                    throw new Error("FETCH messages are not supported");
                }
                else if (msg instanceof FetchOk) {
                    throw new Error("FETCH_OK messages are not supported");
                }
                else if (msg instanceof FetchError) {
                    throw new Error("FETCH_ERROR messages are not supported");
                }
                else if (msg instanceof FetchCancel) {
                    throw new Error("FETCH_CANCEL messages are not supported");
                }
                else if (msg instanceof MaxRequestId) {
                    this.#control.maxRequestId(msg.requestId);
                }
                else if (msg instanceof RequestsBlocked) {
                    console.warn("ignoring REQUESTS_BLOCKED message");
                }
                else if (msg instanceof RequestOk) {
                    // v15: Route RequestOk to both publisher and subscriber
                    await this.#publisher.handleRequestOk(msg);
                    await this.#subscriber.handleRequestOk(msg);
                }
                else if (msg instanceof RequestError) {
                    // v15: Route RequestError to both publisher and subscriber
                    await this.#publisher.handleRequestError(msg);
                    await this.#subscriber.handleRequestError(msg);
                }
                else {
                    unreachable(msg);
                }
            }
            catch (err) {
                console.error("error processing control message", err);
                break;
            }
        }
        console.warn("control stream closed");
    }
    /**
     * Handles a GoAway control message.
     * @param msg - The GoAway message
     */
    async #handleGoAway(msg) {
        console.warn(`MOQLITE_INCOMPATIBLE: Received GOAWAY with redirect URI: ${msg.newSessionUri}`);
        // In moq-lite compatibility mode, we don't support session redirection
        // Just close the connection
        this.close();
    }
    /**
     * Handles an unexpected CLIENT_SETUP control message.
     * @param msg - The CLIENT_SETUP message
     */
    async #handleClientSetup(_msg) {
        console.error("Unexpected CLIENT_SETUP message received after connection established");
        this.close();
    }
    /**
     * Handles an unexpected SERVER_SETUP control message.
     * @param msg - The SERVER_SETUP message
     */
    async #handleServerSetup(_msg) {
        console.error("Unexpected SERVER_SETUP message received after connection established");
        this.close();
    }
    /**
     * Handles object streams (unidirectional streams for media delivery).
     */
    async #runObjectStreams() {
        const readers = new Readers(this.#quic);
        for (;;) {
            const stream = await readers.next();
            if (!stream) {
                break;
            }
            this.#runObjectStream(stream)
                .then(() => {
                stream.stop(new Error("cancel"));
            })
                .catch((err) => {
                console.error("error processing object stream", err);
                stream.stop(err);
            });
        }
    }
    /**
     * Handles a single object stream.
     */
    async #runObjectStream(stream) {
        // we don't support other stream types yet
        const header = await Group.decode(stream);
        await this.#subscriber.handleGroup(header, stream);
    }
    /**
     * Accepts bidirectional streams for v16 SUBSCRIBE_NAMESPACE.
     */
    async #runBidiStreams() {
        for (;;) {
            const stream = await Stream.accept(this.#quic);
            if (!stream)
                break;
            void this.#runBidiStream(stream).catch((err) => {
                console.error("error processing bidi stream", err);
                stream.abort(new Error("bidi stream error"));
            });
        }
    }
    /**
     * Handles a single incoming bidi stream.
     */
    async #runBidiStream(stream) {
        const messageType = await stream.reader.u53();
        if (messageType === SubscribeNamespace.id) {
            await this.#publisher.handleSubscribeNamespaceStream(stream);
        }
        else {
            console.warn(`unexpected bidi stream type: ${messageType}`);
            stream.abort(new Error("unexpected stream type"));
        }
    }
    /**
     * Returns a promise that resolves when the connection is closed.
     * @returns A promise that resolves when closed
     */
    get closed() {
        return this.#quic.closed.then(() => undefined);
    }
}
