// @moq/lite/ietf/fetch.js
import * as Message from "./message.js";
export class Fetch {
    static id = 0x16;
    requestId;
    trackNamespace;
    trackName;
    subscriberPriority;
    groupOrder;
    startGroup;
    startObject;
    endGroup;
    endObject;
    constructor({ requestId, trackNamespace, trackName, subscriberPriority, groupOrder, startGroup, startObject, endGroup, endObject, }) {
        this.requestId = requestId;
        this.trackNamespace = trackNamespace;
        this.trackName = trackName;
        this.subscriberPriority = subscriberPriority;
        this.groupOrder = groupOrder;
        this.startGroup = startGroup;
        this.startObject = startObject;
        this.endGroup = endGroup;
        this.endObject = endObject;
    }
    async #encode(_w) {
        throw new Error("FETCH messages are not supported");
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, Fetch.#decode);
    }
    static async #decode(_r) {
        throw new Error("FETCH messages are not supported");
    }
}
export class FetchOk {
    static id = 0x18;
    requestId;
    constructor({ requestId }) {
        this.requestId = requestId;
    }
    async #encode(_w) {
        throw new Error("FETCH_OK messages are not supported");
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, FetchOk.#decode);
    }
    static async #decode(_r) {
        throw new Error("FETCH_OK messages are not supported");
    }
}
export class FetchError {
    static id = 0x19;
    requestId;
    errorCode;
    reasonPhrase;
    constructor({ requestId, errorCode, reasonPhrase, }) {
        this.requestId = requestId;
        this.errorCode = errorCode;
        this.reasonPhrase = reasonPhrase;
    }
    async #encode(_w) {
        throw new Error("FETCH_ERROR messages are not supported");
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, FetchError.#decode);
    }
    static async #decode(_r) {
        throw new Error("FETCH_ERROR messages are not supported");
    }
}
export class FetchCancel {
    static id = 0x17;
    requestId;
    constructor({ requestId }) {
        this.requestId = requestId;
    }
    async #encode(_w) {
        throw new Error("FETCH_CANCEL messages are not supported");
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, FetchCancel.#decode);
    }
    static async #decode(_r) {
        throw new Error("FETCH_CANCEL messages are not supported");
    }
}
