// @moq/lite/ietf/publish_namespace.js
import * as Message from "./message.js";
import * as Namespace from "./namespace.js";
import { Parameters } from "./parameters.js";
import { Version } from "./version.js";
// In draft-14, ANNOUNCE is renamed to PUBLISH_NAMESPACE
export class PublishNamespace {
    static id = 0x06;
    requestId;
    trackNamespace;
    constructor({ requestId, trackNamespace }) {
        this.requestId = requestId;
        this.trackNamespace = trackNamespace;
    }
    async #encode(w, _version) {
        await w.u62(this.requestId);
        await Namespace.encode(w, this.trackNamespace);
        await w.u53(0); // size of parameters
    }
    async encode(w, version) {
        return Message.encode(w, (wr) => this.#encode(wr, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (rd) => PublishNamespace.#decode(rd, version));
    }
    static async #decode(r, version) {
        const requestId = await r.u62();
        const trackNamespace = await Namespace.decode(r);
        await Parameters.decode(r, version); // ignore parameters
        return new PublishNamespace({ requestId, trackNamespace });
    }
}
export class PublishNamespaceOk {
    static id = 0x07;
    requestId;
    constructor({ requestId }) {
        this.requestId = requestId;
    }
    async #encode(w) {
        await w.u62(this.requestId);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, PublishNamespaceOk.#decode);
    }
    static async #decode(r) {
        const requestId = await r.u62();
        return new PublishNamespaceOk({ requestId });
    }
}
export class PublishNamespaceError {
    static id = 0x08;
    requestId;
    errorCode;
    reasonPhrase;
    constructor({ requestId, errorCode, reasonPhrase, }) {
        this.requestId = requestId;
        this.errorCode = errorCode;
        this.reasonPhrase = reasonPhrase;
    }
    async #encode(w) {
        await w.u62(this.requestId);
        await w.u62(BigInt(this.errorCode));
        await w.string(this.reasonPhrase);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, PublishNamespaceError.#decode);
    }
    static async #decode(r) {
        const requestId = await r.u62();
        const errorCode = Number(await r.u62());
        const reasonPhrase = await r.string();
        return new PublishNamespaceError({ requestId, errorCode, reasonPhrase });
    }
}
export class PublishNamespaceCancel {
    static id = 0x0c;
    trackNamespace;
    requestId; // v16: uses request_id instead of track_namespace
    errorCode;
    reasonPhrase;
    constructor({ trackNamespace = "", errorCode = 0, reasonPhrase = "", requestId = 0n, } = {}) {
        this.trackNamespace = trackNamespace;
        this.requestId = requestId;
        this.errorCode = errorCode;
        this.reasonPhrase = reasonPhrase;
    }
    async #encode(w, version) {
        if (version === Version.DRAFT_16) {
            await w.u62(this.requestId);
        }
        else {
            await Namespace.encode(w, this.trackNamespace);
        }
        await w.u62(BigInt(this.errorCode));
        await w.string(this.reasonPhrase);
    }
    async encode(w, version) {
        return Message.encode(w, (wr) => this.#encode(wr, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (rd) => PublishNamespaceCancel.#decode(rd, version));
    }
    static async #decode(r, version) {
        let trackNamespace = "";
        let requestId = 0n;
        if (version === Version.DRAFT_16) {
            requestId = await r.u62();
        }
        else {
            trackNamespace = await Namespace.decode(r);
        }
        const errorCode = Number(await r.u62());
        const reasonPhrase = await r.string();
        return new PublishNamespaceCancel({ trackNamespace, errorCode, reasonPhrase, requestId });
    }
}
// In draft-14, UNANNOUNCE is renamed to PUBLISH_NAMESPACE_DONE
// In draft-16, uses request_id instead of track_namespace
export class PublishNamespaceDone {
    static id = 0x09;
    trackNamespace;
    requestId; // v16: uses request_id instead of track_namespace
    constructor({ trackNamespace = "", requestId = 0n, } = {}) {
        this.trackNamespace = trackNamespace;
        this.requestId = requestId;
    }
    async #encode(w, version) {
        if (version === Version.DRAFT_16) {
            await w.u62(this.requestId);
        }
        else {
            await Namespace.encode(w, this.trackNamespace);
        }
    }
    async encode(w, version) {
        return Message.encode(w, (wr) => this.#encode(wr, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (rd) => PublishNamespaceDone.#decode(rd, version));
    }
    static async #decode(r, version) {
        if (version === Version.DRAFT_16) {
            const requestId = await r.u62();
            return new PublishNamespaceDone({ requestId });
        }
        const trackNamespace = await Namespace.decode(r);
        return new PublishNamespaceDone({ trackNamespace });
    }
}
