// @moq/lite/ietf/request.js
import * as Message from "./message.js";
import { MessageParameters } from "./parameters.js";
import { Version } from "./version.js";
export class MaxRequestId {
    static id = 0x15;
    requestId;
    constructor({ requestId }) {
        this.requestId = requestId;
    }
    async #encode(w) {
        await w.u62(this.requestId);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async #decode(r) {
        return new MaxRequestId({ requestId: await r.u62() });
    }
    static async decode(r, _version) {
        return Message.decode(r, MaxRequestId.#decode);
    }
}
export class RequestsBlocked {
    static id = 0x1a;
    requestId;
    constructor({ requestId }) {
        this.requestId = requestId;
    }
    async #encode(w) {
        await w.u62(this.requestId);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async #decode(r) {
        return new RequestsBlocked({ requestId: await r.u62() });
    }
    static async decode(r, _version) {
        return Message.decode(r, RequestsBlocked.#decode);
    }
}
/// REQUEST_OK (0x07 in v15) - Generic success response for any request.
/// Replaces PublishNamespaceOk, SubscribeNamespaceOk in v15.
export class RequestOk {
    static id = 0x07;
    requestId;
    parameters;
    constructor({ requestId, parameters = new MessageParameters(), }) {
        this.requestId = requestId;
        this.parameters = parameters;
    }
    async #encode(w, version) {
        await w.u62(this.requestId);
        await this.parameters.encode(w, version);
    }
    async encode(w, version) {
        return Message.encode(w, (wr) => this.#encode(wr, version));
    }
    static async #decode(r, version) {
        const requestId = await r.u62();
        const parameters = await MessageParameters.decode(r, version);
        return new RequestOk({ requestId, parameters });
    }
    static async decode(r, version) {
        return Message.decode(r, (rd) => RequestOk.#decode(rd, version));
    }
}
/// REQUEST_ERROR (0x05 in v15) - Generic error response for any request.
/// Replaces SubscribeError, PublishError, PublishNamespaceError, etc. in v15.
export class RequestError {
    static id = 0x05;
    requestId;
    errorCode;
    reasonPhrase;
    retryInterval;
    constructor({ requestId, errorCode, reasonPhrase, retryInterval = 0n, }) {
        this.requestId = requestId;
        this.errorCode = errorCode;
        this.reasonPhrase = reasonPhrase;
        this.retryInterval = retryInterval;
    }
    async #encode(w, version) {
        await w.u62(this.requestId);
        await w.u62(BigInt(this.errorCode));
        if (version === Version.DRAFT_16) {
            await w.u62(this.retryInterval);
        }
        await w.string(this.reasonPhrase);
    }
    async encode(w, version) {
        return Message.encode(w, (wr) => this.#encode(wr, version));
    }
    static async #decode(r, version) {
        const requestId = await r.u62();
        const errorCode = Number(await r.u62());
        const retryInterval = version === Version.DRAFT_16 ? await r.u62() : 0n;
        const reasonPhrase = await r.string();
        return new RequestError({ requestId, errorCode, reasonPhrase, retryInterval });
    }
    static async decode(r, version) {
        return Message.decode(r, (rd) => RequestError.#decode(rd, version));
    }
}
