// @moq/lite/ietf/subscriber.js
import { Announced } from "../announced.js";
import { Broadcast } from "../broadcast.js";
import { Group } from "../group.js";
import * as Path from "../path.js";
import { Stream } from "../stream.js";
import { error, unreachable } from "../util/error.js";
import { Frame } from "./object.js";
import { PublishError } from "./publish.js";
import { RequestError, RequestOk } from "./request.js";
import { Subscribe, Unsubscribe } from "./subscribe.js";
import { SubscribeNamespace, SubscribeNamespaceEntry, SubscribeNamespaceEntryDone, UnsubscribeNamespace, } from "./subscribe_namespace.js";
import { Version } from "./version.js";
/**
 * Handles subscribing to broadcasts using moq-transport protocol with lite-compatibility restrictions.
 *
 * @internal
 */
export class Subscriber {
    #control;
    // Any currently active announcements.
    #announced = new Set();
    // Any consumers that want each new announcement.
    #announcedConsumers = new Set();
    // Our subscribed tracks - keyed by request ID
    #subscribes = new Map();
    // A map of track aliases to request IDs
    #trackAliases = new Map();
    // Track subscription responses - keyed by request ID
    #subscribeCallbacks = new Map();
    #quic;
    /**
     * Creates a new Subscriber instance.
     * @param control - The control stream writer for sending control messages
     * @param quic - The WebTransport session (needed for v16 bidi streams)
     *
     * @internal
     */
    constructor({ control, quic }) {
        this.#control = control;
        this.#quic = quic;
    }
    /**
     * Gets an announced reader for the specified prefix.
     * @param prefix - The prefix for announcements
     * @returns An AnnounceConsumer instance
     */
    announced(prefix = Path.empty()) {
        const announced = new Announced(prefix);
        for (const active of this.#announced) {
            if (!active.startsWith(prefix))
                continue;
            announced.append({
                path: active,
                active: true,
            });
        }
        this.#announcedConsumers.add(announced);
        this.#runAnnounced(announced, prefix).finally(() => {
            this.#announcedConsumers.delete(announced);
        });
        return announced;
    }
    async #runAnnounced(announced, prefix) {
        if (this.#control.version === Version.DRAFT_16) {
            await this.#runAnnouncedV16(announced, prefix);
        }
        else {
            await this.#runAnnouncedLegacy(announced, prefix);
        }
    }
    async #runAnnouncedLegacy(announced, prefix) {
        const requestId = await this.#control.nextRequestId();
        if (requestId === undefined)
            return;
        try {
            this.#control.write(new SubscribeNamespace({ namespace: prefix, requestId }));
            await announced.closed;
        }
        finally {
            this.#control.write(new UnsubscribeNamespace({ requestId }));
        }
    }
    async #runAnnouncedV16(announced, prefix) {
        const requestId = await this.#control.nextRequestId();
        if (requestId === undefined)
            return;
        const version = this.#control.version;
        try {
            // Open a bidi stream for SUBSCRIBE_NAMESPACE
            const stream = await Stream.open(this.#quic);
            // Write message type + SUBSCRIBE_NAMESPACE
            await stream.writer.u53(SubscribeNamespace.id);
            const msg = new SubscribeNamespace({ namespace: prefix, requestId });
            await msg.encode(stream.writer, version);
            // Read REQUEST_OK or REQUEST_ERROR
            const responseType = await stream.reader.u53();
            if (responseType === RequestOk.id) {
                await RequestOk.decode(stream.reader, version);
            }
            else if (responseType === RequestError.id) {
                const err = await RequestError.decode(stream.reader, version);
                throw new Error(`SUBSCRIBE_NAMESPACE error: code=${err.errorCode} reason=${err.reasonPhrase}`);
            }
            else {
                throw new Error(`unexpected response type: ${responseType}`);
            }
            // Loop reading NAMESPACE / NAMESPACE_DONE messages
            const readLoop = (async () => {
                for (;;) {
                    const done = await stream.reader.done();
                    if (done)
                        break;
                    const msgType = await stream.reader.u53();
                    if (msgType === SubscribeNamespaceEntry.id) {
                        const entry = await SubscribeNamespaceEntry.decode(stream.reader, version);
                        const path = Path.join(prefix, entry.suffix);
                        console.debug(`announced: broadcast=${path} active=true`);
                        this.#announced.add(path);
                        for (const consumer of this.#announcedConsumers) {
                            consumer.append({ path, active: true });
                        }
                    }
                    else if (msgType === SubscribeNamespaceEntryDone.id) {
                        const entry = await SubscribeNamespaceEntryDone.decode(stream.reader, version);
                        const path = Path.join(prefix, entry.suffix);
                        console.debug(`announced: broadcast=${path} active=false`);
                        this.#announced.delete(path);
                        for (const consumer of this.#announcedConsumers) {
                            consumer.append({ path, active: false });
                        }
                    }
                    else {
                        throw new Error(`unexpected message type on subscribe_namespace stream: ${msgType}`);
                    }
                }
            })();
            // Wait for either the read loop to finish or the announced to close
            await Promise.race([readLoop, announced.closed]);
            // Close the bidi stream (replaces UnsubscribeNamespace)
            stream.close();
        }
        catch (err) {
            const e = error(err);
            console.warn(`subscribe_namespace error: ${e.message}`);
        }
    }
    /**
     * Consumes a broadcast from the connection.
     *
     * @param name - The name of the broadcast to consume
     * @returns A Broadcast instance
     */
    consume(path) {
        const broadcast = new Broadcast();
        (async () => {
            for (;;) {
                const request = await broadcast.requested();
                if (!request)
                    break;
                this.#runSubscribe(path, request);
            }
        })();
        return broadcast;
    }
    async #runSubscribe(broadcast, request) {
        const requestId = await this.#control.nextRequestId();
        if (requestId === undefined)
            return;
        this.#subscribes.set(requestId, request.track);
        console.debug(`subscribe start: id=${requestId} broadcast=${broadcast} track=${request.track.name}`);
        const msg = new Subscribe({
            requestId,
            trackNamespace: broadcast,
            trackName: request.track.name,
            subscriberPriority: request.priority,
        });
        // Send SUBSCRIBE message on control stream and wait for response
        const responsePromise = new Promise((resolve, reject) => {
            this.#subscribeCallbacks.set(requestId, { resolve, reject });
        });
        await this.#control.write(msg);
        try {
            const ok = await responsePromise;
            this.#trackAliases.set(ok.trackAlias, requestId);
            console.debug(`subscribe ok: id=${requestId} broadcast=${broadcast} track=${request.track.name}`);
            try {
                await request.track.closed;
                const msg = new Unsubscribe({ requestId });
                await this.#control.write(msg);
                console.debug(`unsubscribe: id=${requestId} broadcast=${broadcast} track=${request.track.name}`);
            }
            finally {
                this.#trackAliases.delete(ok.trackAlias);
            }
        }
        catch (err) {
            const e = error(err);
            request.track.close(e);
            console.warn(`subscribe error: id=${requestId} broadcast=${broadcast} track=${request.track.name} error=${e.message}`);
        }
        finally {
            this.#subscribes.delete(requestId);
            this.#subscribeCallbacks.delete(requestId);
        }
    }
    /**
     * Handles a SUBSCRIBE_OK control message received on the control stream.
     * @param msg - The SUBSCRIBE_OK message
     *
     * @internal
     */
    async handleSubscribeOk(msg) {
        const callback = this.#subscribeCallbacks.get(msg.requestId);
        if (callback) {
            callback.resolve(msg);
        }
        else {
            console.warn("handleSubscribeOk unknown requestId", msg.requestId);
        }
    }
    /**
     * Handles a SUBSCRIBE_ERROR control message received on the control stream.
     * @param msg - The SUBSCRIBE_ERROR message
     *
     * @internal
     */
    async handleSubscribeError(msg) {
        const callback = this.#subscribeCallbacks.get(msg.requestId);
        if (callback) {
            callback.reject(new Error(`SUBSCRIBE_ERROR: code=${msg.errorCode} reason=${msg.reasonPhrase}`));
        }
        else {
            console.warn("handleSubscribeError unknown requestId", msg.requestId);
        }
    }
    /**
     * Handles an ObjectStream message (moq-transport equivalent of moq-lite Group).
     * @param msg - The ObjectStream message
     * @param stream - The stream to read object data from
     *
     * @internal
     */
    async handleGroup(group, stream) {
        const producer = new Group(group.groupId);
        if (group.subGroupId !== 0) {
            throw new Error("subgroups are not supported");
        }
        try {
            let requestId = this.#trackAliases.get(group.trackAlias);
            if (requestId === undefined) {
                // Just hope the track alias is the request ID
                requestId = group.trackAlias;
                console.warn("unknown track alias, using request ID");
            }
            const track = this.#subscribes.get(requestId);
            if (!track) {
                throw new Error(`unknown track: trackAlias=${group.trackAlias} requestId=${this.#trackAliases.get(group.trackAlias)}`);
            }
            // Convert to Group (moq-lite equivalent)
            track.writeGroup(producer);
            // Read objects from the stream until end of group
            for (;;) {
                const done = await Promise.race([stream.done(), producer.closed, track.closed]);
                if (done !== false)
                    break;
                const frame = await Frame.decode(stream, group.flags);
                if (frame.payload === undefined)
                    break;
                // Treat each object payload as a frame
                producer.writeFrame(frame.payload);
            }
            producer.close();
        }
        catch (err) {
            const e = error(err);
            producer.close(e);
            stream.stop(e);
        }
    }
    // we don't support publish, so send PUBLISH_ERROR
    async handlePublish(msg) {
        // TODO technically, we should send PUBLISH_OK if we had a SUBSCRIBE in flight for the same track.
        // Otherwise, the peer will SUBSCRIBE_ERROR because duplicate subscriptions are not allowed :(
        if (this.#control.version === Version.DRAFT_15 || this.#control.version === Version.DRAFT_16) {
            const err = new RequestError({
                requestId: msg.requestId,
                errorCode: 500,
                reasonPhrase: "publish not supported",
            });
            await this.#control.write(err);
        }
        else if (this.#control.version === Version.DRAFT_14) {
            const err = new PublishError({
                requestId: msg.requestId,
                errorCode: 500,
                reasonPhrase: "publish not supported",
            });
            await this.#control.write(err);
        }
        else {
            unreachable(this.#control.version);
        }
    }
    /**
     * Handles a PUBLISH_DONE control message received on the control stream.
     * @param msg - The PUBLISH_DONE message
     */
    async handlePublishDone(msg) {
        // For lite compatibility, we treat this as subscription completion
        const callback = this.#subscribeCallbacks.get(msg.requestId);
        if (callback) {
            callback.reject(new Error(`PUBLISH_DONE: code=${msg.statusCode} reason=${msg.reasonPhrase}`));
        }
    }
    /**
     * Handles a PUBLISH_NAMESPACE control message received on the control stream.
     * @param msg - The PUBLISH_NAMESPACE message
     */
    async handlePublishNamespace(msg) {
        if (this.#announced.has(msg.trackNamespace)) {
            console.warn("duplicate PUBLISH_NAMESPACE message");
            return;
        }
        this.#announced.add(msg.trackNamespace);
        console.debug(`announced: broadcast=${msg.trackNamespace} active=true`);
        for (const consumer of this.#announcedConsumers) {
            consumer.append({
                path: msg.trackNamespace,
                active: true,
            });
        }
    }
    /**
     * Handles a PUBLISH_NAMESPACE_DONE control message received on the control stream.
     * @param msg - The PUBLISH_NAMESPACE_DONE message
     */
    async handlePublishNamespaceDone(msg) {
        if (!this.#announced.has(msg.trackNamespace)) {
            console.warn("unknown PUBLISH_NAMESPACE_DONE message");
            return;
        }
        this.#announced.delete(msg.trackNamespace);
        console.debug(`announced: broadcast=${msg.trackNamespace} active=false`);
        for (const consumer of this.#announcedConsumers) {
            consumer.append({
                path: msg.trackNamespace,
                active: false,
            });
        }
    }
    async handleSubscribeNamespaceOk(_msg) {
        // Don't care
    }
    async handleSubscribeNamespaceError(_msg) {
        throw new Error("SUBSCRIBE_NAMESPACE_ERROR messages are not supported");
    }
    /**
     * Handles a TRACK_STATUS control message received on the control stream.
     * @param msg - The TRACK_STATUS message
     */
    async handleTrackStatus(_msg) {
        throw new Error("TRACK_STATUS messages are not supported");
    }
    // v15: REQUEST_OK replaces SubscribeNamespaceOk, PublishNamespaceOk
    async handleRequestOk(msg) {
        // In v15, RequestOk is used for subscribe namespace acknowledgements
        // Route by request_id — treat similarly to SubscribeNamespaceOk for now
        console.debug("received REQUEST_OK", msg.requestId);
    }
    // v15: REQUEST_ERROR replaces SubscribeNamespaceError, etc.
    async handleRequestError(msg) {
        // In v15, RequestError replaces SubscribeError for subscribe requests
        const callback = this.#subscribeCallbacks.get(msg.requestId);
        if (callback) {
            callback.reject(new Error(`REQUEST_ERROR: code=${msg.errorCode} reason=${msg.reasonPhrase}`));
        }
        else {
            console.warn("handleRequestError unknown requestId", msg.requestId);
        }
    }
}
