// @moq/lite/ietf/track.js
import * as Message from "./message.js";
import * as Namespace from "./namespace.js";
export class TrackStatusRequest {
    static id = 0x0d;
    trackNamespace;
    trackName;
    constructor({ trackNamespace, trackName }) {
        this.trackNamespace = trackNamespace;
        this.trackName = trackName;
    }
    async #encode(w) {
        await Namespace.encode(w, this.trackNamespace);
        await w.string(this.trackName);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, TrackStatusRequest.#decode);
    }
    static async #decode(r) {
        const trackNamespace = await Namespace.decode(r);
        const trackName = await r.string();
        return new TrackStatusRequest({ trackNamespace, trackName });
    }
}
// Track status message for communicating track-level state
export class TrackStatus {
    static id = 0x0e;
    trackNamespace;
    trackName;
    statusCode;
    lastGroupId;
    lastObjectId;
    constructor({ trackNamespace, trackName, statusCode, lastGroupId, lastObjectId, }) {
        this.trackNamespace = trackNamespace;
        this.trackName = trackName;
        this.statusCode = statusCode;
        this.lastGroupId = lastGroupId;
        this.lastObjectId = lastObjectId;
    }
    async #encode(w) {
        await Namespace.encode(w, this.trackNamespace);
        await w.string(this.trackName);
        await w.u62(BigInt(this.statusCode));
        await w.u62(this.lastGroupId);
        await w.u62(this.lastObjectId);
    }
    async encode(w, _version) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, _version) {
        return Message.decode(r, TrackStatus.#decode);
    }
    static async #decode(r) {
        const trackNamespace = await Namespace.decode(r);
        const trackName = await r.string();
        const statusCode = Number(await r.u62());
        const lastGroupId = await r.u62();
        const lastObjectId = await r.u62();
        return new TrackStatus({ trackNamespace, trackName, statusCode, lastGroupId, lastObjectId });
    }
    // Track status codes
    static STATUS_IN_PROGRESS = 0x00;
    static STATUS_NOT_FOUND = 0x01;
    static STATUS_NOT_AUTHORIZED = 0x02;
    static STATUS_ENDED = 0x03;
}
