// @moq/lite/lite/announce.js
import * as Path from "../path.js";
import { unreachable } from "../util/error.js";
import * as Message from "./message.js";
import { Version } from "./version.js";
export class Announce {
    suffix;
    active;
    hops;
    constructor(props) {
        this.suffix = props.suffix;
        this.active = props.active;
        this.hops = props.hops ?? 0;
    }
    async #encode(w, version) {
        await w.bool(this.active);
        await w.string(this.suffix);
        switch (version) {
            case Version.DRAFT_03:
                await w.u53(this.hops);
                break;
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                break;
            default:
                unreachable(version);
        }
    }
    static async #decode(r, version) {
        const active = await r.bool();
        const suffix = Path.from(await r.string());
        let hops = 0;
        switch (version) {
            case Version.DRAFT_03:
                hops = await r.u53();
                break;
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                break;
            default:
                unreachable(version);
        }
        return new Announce({ suffix, active, hops });
    }
    async encode(w, version) {
        return Message.encode(w, (w) => this.#encode(w, version));
    }
    static async decode(r, version) {
        return Message.decode(r, (r) => Announce.#decode(r, version));
    }
    static async decodeMaybe(r, version) {
        return Message.decodeMaybe(r, (r) => Announce.#decode(r, version));
    }
}
export class AnnounceInterest {
    prefix;
    constructor(prefix) {
        this.prefix = prefix;
    }
    async #encode(w) {
        await w.string(this.prefix);
    }
    static async #decode(r) {
        const prefix = Path.from(await r.string());
        return new AnnounceInterest(prefix);
    }
    async encode(w) {
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r) {
        return Message.decode(r, AnnounceInterest.#decode);
    }
}
/// Sent after setup to communicate the initially announced paths.
///
/// Used by Draft01/Draft02 only. Draft03 uses individual Announce messages instead.
export class AnnounceInit {
    suffixes;
    constructor(paths) {
        this.suffixes = paths;
    }
    static #guard(version) {
        switch (version) {
            case Version.DRAFT_01:
            case Version.DRAFT_02:
                break;
            case Version.DRAFT_03:
                throw new Error("announce init not supported for Draft03");
            default:
                unreachable(version);
        }
    }
    async #encode(w) {
        await w.u53(this.suffixes.length);
        for (const path of this.suffixes) {
            await w.string(path);
        }
    }
    static async #decode(r) {
        const count = await r.u53();
        const suffixes = [];
        for (let i = 0; i < count; i++) {
            suffixes.push(Path.from(await r.string()));
        }
        return new AnnounceInit(suffixes);
    }
    async encode(w, version) {
        AnnounceInit.#guard(version);
        return Message.encode(w, this.#encode.bind(this));
    }
    static async decode(r, version) {
        AnnounceInit.#guard(version);
        return Message.decode(r, AnnounceInit.#decode);
    }
}
