// @moq/lite/lite/connection.js
import * as Path from "../path.js";
import { Readers, Stream } from "../stream.js";
import { AnnounceInterest } from "./announce.js";
import { Group } from "./group.js";
import { Publisher } from "./publisher.js";
import { SessionInfo } from "./session.js";
import { StreamId } from "./stream.js";
import { Subscribe } from "./subscribe.js";
import { Subscriber } from "./subscriber.js";
/**
 * Represents a connection to a MoQ server.
 *
 * @public
 */
export class Connection {
    // The URL of the connection.
    url;
    // The version of the connection.
    version;
    // The established WebTransport session.
    #quic;
    // Use to receive/send session messages.
    #session;
    // Module for contributing tracks.
    #publisher;
    // Module for distributing tracks.
    #subscriber;
    // Just to avoid logging when `close()` is called.
    #closed = false;
    /**
     * Creates a new Connection instance.
     * @param url - The URL of the connection
     * @param quic - The WebTransport session
     * @param session - The session stream
     *
     * @internal
     */
    constructor(url, quic, version, session) {
        this.url = url;
        this.#quic = quic;
        this.#session = session;
        this.version = version;
        this.#publisher = new Publisher(this.#quic, this.version);
        this.#subscriber = new Subscriber(this.#quic, this.version);
        this.#run();
    }
    /**
     * Closes the connection.
     */
    close() {
        if (this.#closed)
            return;
        this.#closed = true;
        this.#publisher.close();
        this.#subscriber.close();
        try {
            // TODO: For whatever reason, this try/catch doesn't seem to work..?
            this.#quic.close();
        }
        catch {
            // ignore
        }
    }
    async #run() {
        const session = this.#runSession();
        const bidis = this.#runBidis();
        const unis = this.#runUnis();
        try {
            await Promise.all([session, bidis, unis]);
        }
        catch (err) {
            if (!this.#closed) {
                console.error("fatal error running connection", err);
            }
        }
        finally {
            this.close();
        }
    }
    /**
     * Publishes a broadcast to the connection.
     * @param name - The broadcast path to publish
     * @param broadcast - The broadcast to publish
     */
    publish(path, broadcast) {
        this.#publisher.publish(path, broadcast);
    }
    /**
     * Gets the next announced broadcast.
     */
    announced(prefix = Path.empty()) {
        return this.#subscriber.announced(prefix);
    }
    /**
     * Consumes a broadcast from the connection.
     *
     * @remarks
     * If the broadcast is not found, a "not found" error will be thrown when requesting any tracks.
     *
     * @param broadcast - The path of the broadcast to consume
     * @returns A Broadcast instance
     */
    consume(broadcast) {
        return this.#subscriber.consume(broadcast);
    }
    async #runSession() {
        if (!this.#session) {
            // moq-lite draft-03 doesn't use a session stream.
            return;
        }
        try {
            // Receive messages until the connection is closed.
            for (;;) {
                const msg = await SessionInfo.decodeMaybe(this.#session.reader, this.version);
                if (!msg)
                    break;
                // TODO use the session info
            }
        }
        finally {
            console.debug("session stream closed");
        }
    }
    async #runBidis() {
        for (;;) {
            const stream = await Stream.accept(this.#quic);
            if (!stream) {
                break;
            }
            this.#runBidi(stream)
                .catch((err) => {
                stream.writer.reset(err);
            })
                .finally(() => {
                stream.writer.close();
            });
        }
    }
    async #runBidi(stream) {
        const typ = await stream.reader.u53();
        if (typ === StreamId.Session) {
            throw new Error("duplicate session stream");
        }
        else if (typ === StreamId.Announce) {
            const msg = await AnnounceInterest.decode(stream.reader);
            await this.#publisher.runAnnounce(msg, stream);
            return;
        }
        else if (typ === StreamId.Subscribe) {
            const msg = await Subscribe.decode(stream.reader, this.version);
            await this.#publisher.runSubscribe(msg, stream);
            return;
        }
        else if (typ === StreamId.Probe) {
            await this.#publisher.runProbe(stream);
            return;
        }
        else {
            throw new Error(`unknown stream type: ${typ.toString()}`);
        }
    }
    async #runUnis() {
        const readers = new Readers(this.#quic);
        for (;;) {
            const stream = await readers.next();
            if (!stream) {
                break;
            }
            this.#runUni(stream)
                .then(() => {
                stream.stop(new Error("cancel"));
            })
                .catch((err) => {
                stream.stop(err);
            });
        }
    }
    async #runUni(stream) {
        const typ = await stream.u8();
        if (typ === 0) {
            const msg = await Group.decode(stream);
            await this.#subscriber.runGroup(msg, stream);
        }
        else {
            throw new Error(`unknown stream type: ${typ.toString()}`);
        }
    }
    /**
     * Returns a promise that resolves when the connection is closed.
     * @returns A promise that resolves when closed
     */
    get closed() {
        return this.#quic.closed.then(() => undefined);
    }
}
