// @moq/lite/time.js
export const Nano = {
    zero: 0,
    fromMicro: (us) => (us * 1_000),
    fromMilli: (ms) => (ms * 1_000_000),
    fromSecond: (s) => (s * 1_000_000_000),
    toMicro: (ns) => (ns / 1_000),
    toMilli: (ns) => (ns / 1_000_000),
    toSecond: (ns) => (ns / 1_000_000_000),
    now: () => (performance.now() * 1_000_000),
    add: (a, b) => (a + b),
    sub: (a, b) => (a - b),
    mul: (a, b) => (a * b),
    div: (a, b) => (a / b),
    max: (a, b) => Math.max(a, b),
    min: (a, b) => Math.min(a, b),
};
export const Micro = {
    zero: 0,
    fromNano: (ns) => (ns / 1_000),
    fromMilli: (ms) => (ms * 1_000),
    fromSecond: (s) => (s * 1_000_000),
    toNano: (us) => (us * 1_000),
    toMilli: (us) => (us / 1_000),
    toSecond: (us) => (us / 1_000_000),
    now: () => (performance.now() * 1_000),
    add: (a, b) => (a + b),
    sub: (a, b) => (a - b),
    mul: (a, b) => (a * b),
    div: (a, b) => (a / b),
    max: (a, b) => Math.max(a, b),
    min: (a, b) => Math.min(a, b),
};
export const Milli = {
    zero: 0,
    fromNano: (ns) => (ns / 1_000_000),
    fromMicro: (us) => (us / 1_000),
    fromSecond: (s) => (s * 1_000),
    toNano: (ms) => (ms * 1_000_000),
    toMicro: (ms) => (ms * 1_000),
    toSecond: (ms) => (ms / 1_000),
    now: () => performance.now(),
    add: (a, b) => (a + b),
    sub: (a, b) => (a - b),
    mul: (a, b) => (a * b),
    div: (a, b) => (a / b),
    max: (a, b) => Math.max(a, b),
    min: (a, b) => Math.min(a, b),
};
export const Second = {
    zero: 0,
    fromNano: (ns) => (ns / 1_000_000_000),
    fromMicro: (us) => (us / 1_000_000),
    fromMilli: (ms) => (ms / 1_000),
    toNano: (s) => (s * 1_000_000_000),
    toMicro: (s) => (s * 1_000_000),
    toMilli: (s) => (s * 1_000),
    now: () => (performance.now() / 1_000),
    add: (a, b) => (a + b),
    sub: (a, b) => (a - b),
    mul: (a, b) => (a * b),
    div: (a, b) => (a / b),
    max: (a, b) => Math.max(a, b),
    min: (a, b) => Math.min(a, b),
};
